gc/base/ParallelDispatcher.cpp

```diff
@@ -630,3 +650,134 @@ MM_ParallelDispatcher::reinitAfterFork(MM_EnvironmentBase *env, uintptr_t newThr
+		for (uintptr_t index = newThreadCount; index < _threadCountMaximum; index++) {
+			_statusTable[index] = worker_status_dying;
+		}
+
+		omrthread_monitor_notify_all(_workerThreadMutex);
+
+		omrthread_monitor_exit(_workerThreadMutex);
+
+		uintptr_t expectedThreadShutdownThread = newThreadCount - 1;
+
+		omrthread_monitor_enter(_dispatcherMonitor);
+		while (expectedThreadShutdownThread != _threadShutdownCount) {
+			omrthread_monitor_wait(_dispatcherMonitor);
+		}
+		omrthread_monitor_exit(_dispatcherMonitor);
+
+		/* Cleanup the dispatcher tables */
+		for(uintptr_t index = newThreadCount; index < _threadCountMaximum; index++) {
+			Assert_MM_true(worker_status_dying == _statusTable[index]);
+			_statusTable[index] = worker_status_inactive;
+			_threadTable[index] = NULL;
+		}
+
+		Assert_MM_true(_threadShutdownCount == expectedThreadShutdownThread);
+
+		_extensions->gcThreadCount = newThreadCount;
+		_activeThreadCount = newThreadCount;
+		_threadCount = newThreadCount;
+		_threadCountMaximum = newThreadCount;
+
+		_inShutdown = false;
+
+		Trc_MM_ParallelDispatcher_criu_contractThreadPool_Success(preShutdownThreadCount, newThreadCount);
+	}
+
+	Trc_MM_ParallelDispatcher_criu_contractThreadPool_Exit(_extensions->gcThreadCount);
+}
+
+/**
+ * Expand/fill the thread pool by starting up threads based on what H/W suppourts. This API is capped by the initial
+ * thread pool size, i.e we can't expand past _threadCountMaximumAtCheckpointStartup (currently don't support expanding dispatcher tables).
+ *
+ * This API assumes  CRIU is the only consumer (not tested for general use).
+ * The following conditions are required while expanding the thread pool:
+ *     1) NOT holding exclusive VM access
+ *     2) Dispatcher is idle (no task can be dispatched)
+ *     3) Can't be in/enter shutdown
+ */
+bool
+MM_ParallelDispatcher::expandThreadPool(MM_EnvironmentBase *env)
+{
+	Trc_MM_ParallelDispatcher_criu_expandThreadPool_Entry();
+
+	Assert_MM_false(_workerThreadsReservedForGC);
+	Assert_MM_false(_inShutdown);
+
+	Assert_MM_true(_threadShutdownCount == (_threadCountMaximum - 1));
+	Assert_MM_true(_threadCountMaximum == _extensions->gcThreadCount);
+
+	bool result = true;
+
+	uintptr_t supportedGCThreadCount = _extensions->configuration->supportedGCThreadCount(env);
+	uintptr_t preRestoreThreadCount = _extensions->gcThreadCount;
+	uintptr_t newThreadCount = OMR_MIN(_threadCountMaximumAtCheckpointStartup, supportedGCThreadCount);
+
+	Trc_MM_ParallelDispatcher_criu_expandThreadPool_params(newThreadCount, _threadCountMaximumAtCheckpointStartup, supportedGCThreadCount, preRestoreThreadCount);
+
+	if (newThreadCount > preRestoreThreadCount) {
+
+		result = internalStartupThreads(preRestoreThreadCount, newThreadCount);
+
+		if (result) {
+			Trc_MM_ParallelDispatcher_criu_expandThreadPool_Success();
+			Assert_MM_true(_threadShutdownCount == (newThreadCount - 1));
+		} else {
+			Trc_MM_ParallelDispatcher_criu_expandThreadPool_Failed(_threadShutdownCount);
```

Comment:
you could move those 2 trace points (and actually have only one reporting result and whatever relevant internals you want) inside internalStartupThreads. it would be interesting in non-restore startup potential (albeit far less likely) failure, as well

I'd drop _criu_ from the name of trace points

Reply:
Made the suggested changes along with few other things after revisiting trace points:

- Removed "criu" from trc point names/template descriptions
- Changed Trc point group identifier from "criudispatcher" to "dispatcher"
- Removed trace points:
`Trc_MM_ParallelDispatcher_criu_expandThreadPool_Failed` and
`Trc_MM_ParallelDispatcher_criu_expandThreadPool_Success`.
- Introduced new trace point `Trc_MM_ParallelDispatcher_internalStartupThreads_Failed` to log start up thread failure.
- Cleaned up template descriptions for consistency (Entry/Exit trace points description aligned for both checkpoint and restore trc points)
- Renamed `preRestoreThreadCount` variable to `preExpandThreadCount`